experimental/beacon_sim/analysis/belief_propagation.py

```diff
@@ -25,46 +25,67 @@ def remap_vars(self):
         M = self.model.H().transpose() @ inv(self.model.R) @ self.model.H()
         return G, R, M
 
-    def predict_covariance(self,P):
-        P = self.model.F() @ P @ self.model.F().transpose() + self.model.L() @ self.model.Q @ self.model.L().transpose()
+    def predict_covariance(self,P,backward=False):
+        if not backward:
+            P = self.model.F() @ P @ self.model.F().transpose() + self.model.L() @ self.model.Q @ self.model.L().transpose()
+        else:
+            P = inv(self.model.F()) @ (P - self.model.L() @ self.model.Q @ self.model.L().transpose()) @ inv(self.model.F().transpose())
         return P
 
-    def update_covariance(self,P):
-        n = P.shape[0]
-        S = self.model.H() @ P @ self.model.H().transpose() + self.model.M() @ self.model.R @ self.model.M().transpose()
-        K= P @ self.model.H().transpose() @ inv(S)
-        return (np.identity(n) - K @ self.model.H()) @ P
+    def update_covariance(self,P,backward=False):
+        if not backward:
+            n = P.shape[0]
+            S = self.model.H() @ P @ self.model.H().transpose() + self.model.M() @ self.model.R @ self.model.M().transpose()
+            K= P @ self.model.H().transpose() @ inv(S)
+            return (np.identity(n) - K @ self.model.H()) @ P
+        else:
+            n = P.shape[0]
+            Sbar = self.model.M() @ self.model.R @ self.model.M().transpose() - self.model.H() @ P @ self.model.H().transpose()
+            Kbar = P @ self.model.H().transpose() @ inv(Sbar)
+            return (np.identity(n) + Kbar @ self.model.H()) @ P
```

Comment:
Is this the same as $\Sigma_{t+1|t} = (\Sigma_{t+1}^{-1} - H^T R^{-1}H)^{-1}$?

Reply:
I am just seeing this now. Yes, but rearranged applying the matrix inversion lemma to something that looks like the standard form.

Reply:
(this form only requires 1 matrix inverse)